Plugin.pm

```diff
@@ -521,16 +521,23 @@ sub QobuzArtist {
 			passthrough => [{
 				artistId  => $artist->{id},
 			}],
-		},{
-			name => cstring($client, 'PLUGIN_QOBUZ_ADD_FAVORITE', $artist->{name}),
-			image => 'html/images/favorites.png',
-			url  => \&QobuzAddFavorite,
-			passthrough => [{
-				artist_ids => $artist->{id},
-			}],
-			nextWindow => 'parent'
 		};
-;
+		
+		Plugins::Qobuz::API->getUserFavorites(sub {
+			my $favorites = shift;
+			my $artistId = $artist->{id};
+			my $isFavorite = grep { $_->{id} eq $artistId } @{$favorites->{artists}->{items}};
+		
+			push @$items, {
+ 				name => cstring($client, $isFavorite ? 'PLUGIN_QOBUZ_REMOVE_FAVORITE' : 'PLUGIN_QOBUZ_ADD_FAVORITE', $artist->{name}),
+				url  => $isFavorite ? \&QobuzDeleteFavorite : \&QobuzAddFavorite,
+				image => 'html/images/favorites.png',
+				passthrough => [{
+					artist_ids => $artist->{id},
+				}],
+				nextWindow => 'parent'
+			}
+		});
```

Comment:
Unfortunately that wouldn't work (reliably): as the call might need to go online, and that lookup is asynchronous, the addition of the menu item might happen _after_ the menu has been returned to the player. You'd have to queue up these async requests and only return the menu once all the requests have been handled... welcome in callback hell!

What I've been doing in some places is create a small helper function which would return the menu (the final `$cb->()` call). That helper function would know the list of async functions. And only once all of them had responded it would actually call the `$cb->()`. 

Or, if you didn't mind making this available to LMS 8+ users only, you could use [`Async::Util::achain()`](https://metacpan.org/pod/Async::Util#achain) to simplify this. I only added this module in 8.0 (I think), in support of the Online Music Library Integration.

One more thing to keep in mind with async functions: they don't always return in the order you fired them off. Therefore `push` won't work. You'd have to somehow tell the callback at what position to insert the item (something like `splice @$items, $position, 0, $newItem` - see https://perldoc.perl.org/functions/splice).

Reply:
Drats! I knew that was too easy. ;)  Thanks for the feedback. I'm sure I'll have more questions as I look into it further. As for requiring LMS 8+, I would have no problem with it. Could it be done with conditional code or would the plugin only be available to LMS 8+ users?

Reply:
`if (Slim::Utils::Versions->compareVersions($::VERSION, '8.0.0') >= 0) { ... }` - see at the top of that file.